decisions/0003-package-for-shared-server-features.md

```diff
@@ -0,0 +1,113 @@
+# Package for shared server features
+
+Date: 2022-06-01
+
+Status: proposed
+
+## Context
+
+Currently, users need to install 4 Remix dependencies:
+
+- Dev (`@remix-run/dev`)
+- Server runtime (`@remix-run/{cloudflare,deno,node}`)
+- Adapter (`@remix-run/{arc,cloudflare-pages,cloudflare-workers,express,fly,netlify,vercel}`)
+- Renderer (`@remix-run/react`)
+
+User code currently looks like this:
+```tsx
+// server runtime
+import { json, createFileSessionStorage } from "@remix-run/node";
+// adapter
+import { createRequestHandler } from "@remix-run/express"
+// renderer
+import { useLoaderData } from "@remix-run/react";
+```
+
+### Goals
+
+- make DX and docs simpler: more intuitive, less magic, eliminate some user errors
+- reduce bugs from `remix` internal complexity
+
+### Non-goals
+
+- maintainability of `remix` repo
+- more portable `app/` directory
+- reduce or increase number of packages
+
+Some of these are nice-to-haves, but not the primary reasons for this proposal.
+
+### Issues
+
+1. Users can import features from their runtime that aren't supported by their adapter:
+
+```tsx
+import { createFileSessionStorage } from "@remix-run/node";
+// Implementation only depends on server runtime, but _availability_ depends on the adapter!
+// ❌ This will crash in netlify nor in vercel!
+```
+
+2. Docs, examples, and videos refer to `@remix-run/node` by default for shared functionality like `json`:
+
+```tsx
+import { json } from "@remix-run/node";
+// Need docs to tell users that they should import from their corresponding runtime package, 
+// and not from `@remix-run/server-runtime`.
+```
+
+3. Keeping shared exports consistent across runtime packages is [error-prone](https://github.com/remix-run/remix/pull/3215) and tedious.
+
+## Decision
+
+Users should import features from _adapter packages_, not from runtime packages. Each adapter can export only the features it supports. Shared features will be imported from a new package, `@remix-run/server`.
+
+This means there won't be any imports needed from the runtime packages (`@remix-run/{cloudflare,deno,node`}) and these will become an implementation detail for most users. Only developers that need to develop adapters will import from the runtime packages directly.
+
+### User-facing packages
+
+Users still need to import from 4 Remix packages:
+
+- Dev `@remix-run/dev` (unchanged)
+- Shared server features `@remix-run/server`
+- Adapter-specific server features (`@remix-run/{arc,cloudflare-pages,cloudflare-workers,express,fly,netlify,vercel}`)
```

Comment:
What would be the recommended adapter when just figuring things out locally?

Right now if you're using the `fly` or `remix` template, you don't have an adapter installed and are just importing functions from `@remix-run/node`.
Since we want people to import as much as possible from an adapter, should we create a `remix` adapter?
Or do we want to recommend another adapter (like `@remix-run/fly`, which currently also doesn't exist) instead?

The `deno` template has the same problem. We however have another concern here as `@remix-run/deno` is the runtime, so should we create a `@remix-run/deno-deploy` adapter for this?

Reply:
The `remix` template uses `@remix-run/serve`, and consequently, `@remix-run/express`. So it should either (a) re-export adapter-specific features from `@remix-run/express` or (b) declare `@remix-run/express` as a peer dependency and users can import directly from `@remix-run/express` for adapter-specific features. 

---

I think it does make sense to have a `@remix-run/fly` adapter that only exports the features available on Fly.

---

For Deno: yes, I think we should create a `@remix-run/deno-deploy` adapter.

Reply:
`@remix-run/serve` isn't really an adapter, but a way to serve the code, right?

It's used in both `fly` & `remix` templates to do so.

When using `@remix-run/netlify` adapter, we also need an extra package (`netlify-cli` globally installed) to serve the code, so it's also not coming from the adapter directly.
Same when using `@remix-run/cloudflare-pages`, we use `wrangler` to serve the code.

So I think we still need an adapter package for the `remix` template (next to the new `fly` adapter) 🤔

Reply:
> @remix-run/serve isn't really an adapter, but a way to serve the code, right?

That's right. The revelant adapter for `@remix-run/serve` and the `remix` template is `@remix-run/express`. That's why I'm leaning towards having `@remix-run/serve` specify `@remix-run/express` as a peer dependency, so that users can still import adapter-specific features from `@remix-run/express`.

Reply:
`@remix-run/serve` is also used in the `fly` & `vercel` template, so especially for the `vercel` template we can't use `@remix-run/express`.

Reply:
Yea there's really a conflation between two things:
1) The request/response interface, which is specific to e.g. express
2) The server features available for a host, which is specific to e.g fly or vercel or etc...

I'll think a bit about what we can do to capture this distinction more clearly and what the API for that would be, i.e. where to import what.